File: cpp/OddEvenLinkedList.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* oddEvenList(ListNode* head) {
    if(head==NULL || head->next==NULL) return head;
    ListNode* p = head, *q = head->next, *even=head->next;
    while(q->next&&p->next){
        p->next = q->next;
        p=p->next;
        if(!p->next) break;
        q->next=p->next;
        q=q->next;
    }
    q->next = NULL;
    p->next = even;
    return head;
}

File: java/Linkedlist/OddEvenLinkedList.java
package Linkedlist;

// https://leetcode.cn/problems/odd-even-linked-list/
public class OddEvenLinkedList {
    public ListNode oddEvenList(ListNode head) {
        if (head == null) return null;
        ListNode evenhead = head.next;
        ListNode even = evenhead, odd = head;
        while (even != null && even.next != null) {
            odd.next = even.next;
            odd = odd.next;
            even.next = odd.next;
            even = even.next;
        }
        odd.next = evenhead;
        return head;
    }
}
